src/rollup/widgets/actions/construct-widget-actions.js

```diff
@@ -209,48 +176,92 @@ export default function constructWidgetActions (widget, config) {
+
+      if (typeof widgetFormValidate === 'function') {
+        switch (config.type) {
+        case 'form':
+          widgetAction = 'submit';
+          break;
+        case 'subscription':
+          widgetAction = 'subscribe';
+          break;
+        case 'sitegate':
+          widgetAction = 'unlock';
+          break;
+        }
 
-          if (typeof config.confirmAction.callback === 'function') {
-            config.confirmAction.callback(callbackTypes.MODAL_CONFIRM, {
+        if (!widgetAction || !widgetFormValidate(event)) {
+          return;
+        } else if (widgetForm) {
+          trackWidgetAction(widgetAction, config, widgetForm);
+
+          data = Array.prototype.slice.call(
+            widgetForm.querySelectorAll('input, textarea, select')
+          ).map(function (element) {
+            return {
+              name: element.name || element.id,
+              value: element.value
+            };
+          });
+
+          // onSubmit callback should be deprecated,
+          // we keep the cb for backwards compatibility.
+          if (typeof config.onSubmit === 'function') {
+            config.onSubmit(callbackTypes.FORM_SUBMIT, {
               widget: widget,
               config: config,
-              event: event
+              event: event,
+              data: data
             });
           }
         }
+      }
 
-        if (typeof widgetOnButtonClick === 'function') {
-          widgetOnButtonClick(event);
+      trackWidgetAction('confirm', config);
+      updateActionCookie(PREFIX_CONFIRM + widget.id, config.expiration);
+
+      if (typeof config.confirmAction === 'object') {
+        if (config.confirmAction.close === false) {
+          shouldClose = false;
         }
 
-        if (shouldClose) {
-          if (config.layout !== 'inline' && typeof config.success === 'undefined') {
-            closeWidget(widget.id, true);
-            widgetOnModalClose(widget, config, event);
+        if (typeof config.confirmAction.callback === 'function') {
+          var param = {
+            widget: widget,
+            config: config,
+            event: event
+          };
 
-          // show success state
-          } else {
-            addClass(widget, 'success');
+          if (data) {
+            param.data = data;
+          }
 
-            // default to a three second delay if the user has not defined one
-            var delay = typeof config.success.delay !== 'undefined' ? config.success.delay * 1000 : 3000;
+          config.confirmAction.callback(callbackTypes.MODAL_CONFIRM, param);
+        }
+      }
 
-            if (delay > 0) {
-              setTimeout(function () {
-                closeWidget(widget.id, true);
-              }, delay);
-            }
+      if (typeof widgetOnButtonClick === 'function') {
+        widgetOnButtonClick(event);
+      }
+
+      if (shouldClose) {
+        if (config.layout !== 'inline' && typeof config.success === 'undefined') {
```

Comment:
I find it useful to do `config.success == null` vs `typeof config.success === 'undefined'`, but I see that it's a pattern in this code base. Up to you.